### PR TITLE
Update DAGs to not download intermediate results by default.

### PR DESCRIPTION
**Pull request by thetorpedodog**
This enables callers to not actually download the results of a
server-side call with a new `_download_results` parameter to each of the
`whatever_base` functions. If set, it will tell the server not to send
the results of the function when it is complete; just to return headers.
(This is useful even *before* the server has support for not sending
results back—we will get the HTTP response and completely ignore
the contents, meaning that the dataset will not be parsed into memory.)

This requires restructuring of the Node class itself, but maintains
(much of) the same external interface.

The actual future used by the instance internally is privatized and now
wraps a `Result` rather than the object that the UDF call would have
produced. The (not-quite official, but still present) external-facing
`.future` is now a property that materializes a proxy `Future` that
receives the result of the actual Future.

---

There is some weirdness in the way we handle concurrency and cancellation right now but those are no worse than the existing implementation, just slightly different.

### CODE REVIEW

**Comment by thetorpedodog** [comment]
Bump on this—this is a bug we’ve had open for a while.

File: tiledb/cloud/_results/results.py
Comment (ihnorton) [inline tiledb/cloud/_results/results.py:61]: Do we need to keep `_body` after decoding?

File: tiledb/cloud/dag/dag.py
Comment (ihnorton) [inline tiledb/cloud/dag/dag.py:175]: Based on the comment above, don't we need a deep copy here?

File: tiledb/cloud/client.py
Comment (ihnorton) [inline tiledb/cloud/client.py:500]: TIL `thread_name_prefix` 👍

File: tiledb/cloud/_results/results.py
Comment (ihnorton) [inline tiledb/cloud/_results/results.py:81]: Mostly out of curiosity, why are you using `.ib` rather than `field` + type annotation here?

File: tiledb/cloud/_results/results.py
Comment (ihnorton) [inline tiledb/cloud/_results/results.py:101]: Is the validator called (1) only in the constructor and (2) after `_body` is actually set on the object?

**Review by ihnorton — COMMENTED** [review]
AFAICT this is using attrs for two classes, providing:
- automatic `__init__` (also available w/ dataclasses IIUC)
- automatic `__eq__` (not available w/ dataclasses)
- automatic execution of `_body` validator (not available w/ dataclasses)

Are there ways it is implicitly used?

The dependency is fine with me, but please continue to be judicious about dependencies here.

File: tiledb/cloud/_results/results.py
Comment (thetorpedodog) [reply tiledb/cloud/_results/results.py:61]: We do not! Good idea.

File: tiledb/cloud/_results/results.py
Comment (thetorpedodog) [reply tiledb/cloud/_results/results.py]: Mostly because at the time, `attr.ib` was the primary API for attrs. I switched it over to the modern version.

File: tiledb/cloud/_results/results.py
Comment (thetorpedodog) [reply tiledb/cloud/_results/results.py:104]: This is called only at initialization time.

File: tiledb/cloud/dag/dag.py
Comment (thetorpedodog) [reply tiledb/cloud/dag/dag.py:175]: In this case, no; our concerns are only about the arguments we add to the dictionary (`_download_results` and `stored_param_uuids`). I updated the comment to be a bit more precise here.

**Review by thetorpedodog — COMMENTED** [review]
Good questions!

> AFAICT this is using attrs for two classes, providing:
>
> - automatic __init__ (also available w/ dataclasses IIUC)
> - automatic __eq__ (not available w/ dataclasses)
> - automatic execution of _body validator (not available w/ dataclasses)
>
> Are there ways it is implicitly used?
> 
> The dependency is fine with me, but please continue to be judicious about dependencies here.

Yes, those are the ways we use attrs; it's all explicit. One of the considerations I used when deciding to make this switch was that at least one of our dependencies already itself depends upon attrs, so directly depending upon it does not actually increase our net dependency count.

File: tiledb/cloud/_results/results.py
Comment (ihnorton) [reply tiledb/cloud/_results/results.py:104]: What I'm wondering is when `_body` is set here -- does attrs set it first _and then_ call the validator?

File: tiledb/cloud/dag/dag.py
Comment (ihnorton) [reply tiledb/cloud/dag/dag.py:175]: Got it, ok, thanks.

File: tiledb/cloud/_results/results.py
Comment (thetorpedodog) [reply tiledb/cloud/_results/results.py:104]: Oh, yes. Validators are run only after the rest of the `__init__` method has completed, so at that point the class is fully set up, and validation ensures that the setup is correct before it is returned to the caller.

File: tiledb/cloud/_results/results.py
Comment (ihnorton) [reply tiledb/cloud/_results/results.py]: Alright, your call

File: tiledb/cloud/_results/results.py
Comment (ihnorton) [reply tiledb/cloud/_results/results.py:104]: Thanks

**Review by ihnorton — APPROVED** [review]
Approving, but please give the tests a once-over and see if there's anywhere else we can add coverage here. It doesn't _seem_ like the scope of changes here is fully reflected in the changes to `test_dag.py`. But that might be due to sufficient coverage already?

**Comment by thetorpedodog** [comment]
> Approving, but please give the tests a once-over and see if there's anywhere else we can add coverage here. It doesn't _seem_ like the scope of changes here is fully reflected in the changes to `test_dag.py`. But that might be due to sufficient coverage already?

We have, to some degree, already covered this with our tests of running server-side, but nothing specifically tested what happened when we tried not to download. Added a test for this.

Thanks for the great review!

**Pull request merged**
